### ISSUE TITLE
Embed rewriting not perfect

### ISSUE THREAD

**Issue by keith-hall** [trishume/syntect#222]
After the latest fixes, I tried updating the submodules, and can see we still have some syntax test failures, this time relating to `embed`...`escape`:

```
FAILED testdata/Packages\Java\syntax_test_java.java: 817
FAILED testdata/Packages\Markdown\syntax_test_markdown.md: 61
```

We get the same failures in Sublime Text, when rewriting the `embed`...`escape` rules to `with_prototype` like `syntect`'s `yaml_load` is doing, so the good news is that `syntect`s parser is working correctly:

`JavaDoc.sublime-syntax`:
```diff
--- Shipped Packages/Java/JavaDoc.sublime-syntax    2018-10-11 19:11:54
+++ Packages/Java/JavaDoc.sublime-syntax    2018-10-30 11:28:20
@@ -20,11 +20,12 @@
     - meta_include_prototype: false
     - match: /\*\*
       scope: comment.block.documentation.javadoc punctuation.definition.comment.begin.javadoc
-      embed: contents
-      embed_scope: comment.block.documentation.javadoc text.html.javadoc
-      escape: \*/
-      escape_captures:
-        0: comment.block.documentation.javadoc punctuation.definition.comment.end.javadoc
+      push:
+        - [{ meta_include_prototype: false }, { meta_content_scope: 'comment.block.documentation.javadoc text.html.javadoc' }, { match: '\*/', scope: comment.block.documentation.javadoc punctuation.definition.comment.end.javadoc, pop: true }]
+        - contents
+      with_prototype:
+        - match: (?=\*/)
+          pop: true
```

`Markdown.sublime-syntax`:
```diff
--- Shipped Packages/Markdown/Markdown.sublime-syntax   2018-10-11 19:11:54
+++ Packages/Markdown/Markdown.sublime-syntax   2018-10-30 09:44:46
@@ -1106,12 +1106,12 @@ contexts:
         0: meta.code-fence.definition.begin.html-php.markdown-gfm
         2: punctuation.definition.raw.code-fence.begin.markdown
         5: constant.other.language-name.markdown
-      embed: scope:embedding.php
-      embed_scope: markup.raw.code-fence.html-php.markdown-gfm
-      escape: '{{code_fence_escape}}'
-      escape_captures:
-        0: meta.code-fence.definition.end.html-php.markdown-gfm
-        1: punctuation.definition.raw.code-fence.end.markdown
+      push:
+        - [{ meta_include_prototype: false }, { meta_content_scope: 'markup.raw.code-fence.html-php.markdown-gfm' }, { match: '{{code_fence_escape}}', captures: { 0: meta.code-fence.definition.end.html-php.markdown-gfm, 1: punctuation.definition.raw.code-fence.end.markdown }, pop: true }]
+        - scope:embedding.php
+      with_prototype:
+        - match: (?={{code_fence_escape}})
+          pop: true
     - match: |-
          (?x)
           {{fenced_code_block_start}}
```

For Java(Doc), the workaround suggested at https://github.com/trishume/syntect/issues/160#issuecomment-388378965 works.

For Markdown, the following syntax test is enough to see the problem:

`````markdown
| SYNTAX TEST "Packages/Markdown/Markdown.sublime-syntax"
```html+php
<div></div>
|^^^ entity.name.tag.block.any.html
<?php
|^^^^ punctuation.section.embedded.begin.php
var_dump(expression);
| ^^^^^^ support.function.var.php
```
|^^ punctuation.definition.raw.code-fence.end.markdown
`````

Everything except the last line's assertions pass. Adding `?>` before the closing backticks/code fence allows the backticks to be correctly recognized as the closing code fence instead of PHP interpolated string punctuation, but we will need a solution so that the way `syntect` handles `embed` actions will work 100% correctly in all situations.

**Comment by keith-hall** [comment]
I guess the fundamental flaw with rewriting `embed` actions to the "equivalent" `with_prototype` construct is that `embed` is guaranteed to escape at the `escape` pattern, but with `with_prototype`, a match pattern could consume the characters first.

Simple example `embed_vs_with_prototype.sublime-syntax`:
```yaml
%YAML 1.2
---
# See http://www.sublimetext.com/docs/3/syntax.html
scope: source.embed-vs-with_prototype
contexts:
  main:
    - match: embed
      scope: keyword.begin
      embed: contents
      escape: escape
      escape_captures:
        0: keyword.end
    - match: with_prototype
      scope: keyword.begin
      push: 
        - [{ meta_include_prototype: false }, { match: escape, scope: keyword.end, pop: true }]
        - contents
      with_prototype:
        - match: (?=escape)
          pop: true
  contents:
    - match: \w+
      scope: string
```

Test text:
```
embedtesting123escape
with_prototypetesting123escape
```
![image](https://user-images.githubusercontent.com/11882719/47775522-91263300-dcf8-11e8-9511-44aada4930cf.png)

So unfortunately, I believe we will need to include native support for `embed`/`escape` in the parser.

**Comment by robinst** [comment]
Makes sense! So escape needs to have precedence over other matches.

**Comment by keith-hall** [comment]
Just a note for when we work on this, there is an [open question for SublimeHQ to answer](https://github.com/SublimeTextIssues/Core/issues/2342#issuecomment-440943499) regarding whether the `prototype` context should be ignored in contexts `embed`ded from the `prototype`. Currently (as at build 3180) Sublime Text's behavior is to apply the `prototype`.

**Comment by robinst** [comment]
@keith-hall So let me get this straight, in this example:

```
embedtesting123escape
```

`embed` matches the embed, that's easy. But then the next match would be that `\w+` matches all of `testing123escape` with scope `string`. *But* the embed escape also matches `escape`.

So in that case, the escape match takes precedence and the match for `string` is cut short.

I'm wondering how that's implemented (and how we would implement it). Does something like this make sense?:

1. If we are currently in an embed, try matching the escape first. If that matches:
   * if the match is at our current position, pop out of the embed
   * else continue matching other contexts but limit it to text that comes before the match (basically truncate our input).
2. Advance position to the next match and repeat

What happens if there are multiple active embeds that has matching escapes, which one wins?

**Comment by keith-hall** [comment]
yes, that makes sense, and sounds like the way forward. 
For multiple embeds, I believe the same ordering as `with_prototype` is used - the embedded contexts are prioritized by the order in which were pushed onto the stack, an earlier embed's `escape` wins even if a later escape would match at an earlier index in the string, if that makes sense. It may be better illustrated with an example, I will try to come up with one when I get a chance :)

**Comment by robinst** [comment]
I tried this example:

```
%YAML 1.2
---
# See http://www.sublimetext.com/docs/3/syntax.html
scope: source.nested-embed
contexts:
  main:
    - match: \(
      scope: paren.begin
      embed: a
      escape: \)
      escape_captures:
        0: paren.end
  a:
    - match: \w+
      scope: string.a
    - match: \<
      scope: bracket.begin
      embed: b
      escape: \>
      escape_captures:
        0: bracket.end
  b:
    - match: \w+
      scope: string.b
```

With the input `( < > )`, the `>` is `bracket.end` and `)` is `paren.end`.

With `( < ) >`, `)` is `paren.end` and `)` doesn't have a scope. So yeah, the earlier embed's escape wins and pops out all the way. (I also did an example where I used the same pattern for both escapes, and it popped out all the way as well.)

**Comment by keith-hall** [comment]
nice :+1: one other thing I wanted to check was how escapes are handled if one escape pattern matches inside another one... say, `escape: me` and `escape: test-me` with text `test-me` - I imagine the first embed should be popped as it should win with `me`. Sorry for not doing the due diligence myself or laying out the yaml for a proper test case, I'm on mobile atm.

**Comment by robinst** [comment]
Just tested, `me` wins (if it's the earlier embed).

**Comment by stefanobaghino** [comment]
FYI I have a patch ready for this ([diff](https://github.com/stefanobaghino/syntect/compare/branch-point...native-embed-escape?expand=1)) ready to be turned into a PR once #614 is merged.

I also went a bit further and bumped the packages on top of that to see what the next steps are and I think we might be in a position where we can cleanly move everything syntax tests from `st2` to `master` without problems.

To the best of my understanding the changes made so far require a new major release. If that matches your understanding as well, at some point I'd like to have a conversation with regard to whether there are other issues that we'd like to tackle before the major release to batch all breaking changes together. I would imagine all relevant tickets can be put into https://github.com/trishume/syntect/milestone/1 to that end.

**Comment by jrappen** [comment]
> ... `st2` to `master` ...

Just for clarification ....

* `embed` was added in Sublime Text Build 3153, compare <https://www.sublimetext.com/docs/syntax.html#match-patterns>.
* It's [`st3`](https://github.com/sublimehq/Packages/tree/st3/) to [`master`](https://github.com/sublimehq/Packages/) for the [`.gitmodules` of `syntect`](https://github.com/trishume/syntect/blob/master/.gitmodules), but you knew that. There is no `st2` branch.
  Compare the discussion in #271 and the follow-up PR in #279 where the `.gitmodules` target for `Packages` was temporarily (and still is) set to `st3`.
* As @keith-hall already said, not everything works for `syntect` using `st3`, though. Specifically this issue, not sure if there was something else, too. `st3` branch of the Packages repo targets the last Builds on the `stable` and `dev` release channels of Sublime Text, which are the same if I recall correctly but just published to both channels for organizational reasons.

**Comment by stefanobaghino** [comment]
@jrappen

> * `embed` was added in Sublime Text Build 3153, compare https://www.sublimetext.com/docs/syntax.html#match-patterns.
> * It's [`st3`](https://github.com/sublimehq/Packages/tree/st3/) to [`master`](https://github.com/sublimehq/Packages/) for the [`.gitmodules` of `syntect`](https://github.com/trishume/syntect/blob/master/.gitmodules), but you knew that. There is no `st2` branch.
>   Compare the discussion in [sublime-syntax branch_point support #271](https://github.com/trishume/syntect/issues/271) and the follow-up PR in [Packages submodule, git files, and docs improvements #279](https://github.com/trishume/syntect/pull/279) where the `.gitmodules` target for `Packages` was temporarily (and still is) set to `st3`.

Thanks for catching my typo and provide additional context. 🙂 I'll have a better look at the pointers you provided, thanks for those as well.

> * As [@keith-hall](https://github.com/keith-hall) already said, not everything works for `syntect` using `st3`, though. Specifically this issue, not sure if there was something else, too. `st3` branch of the Packages repo targets the last Builds on the `stable` and `dev` release channels of Sublime Text, which are the same if I recall correctly but just published to both channels for organizational reasons.

That matches my understanding as well. IIUC that would be the point I captured in [this comment](https://github.com/trishume/syntect/pull/615#issuecomment-4153422356) with regard to existing known failures.
